File: src/pearl.py
import copy
import sys
import math
from collections import defaultdict, deque

import numpy as np
from sklearn.metrics import cohen_kappa_score
from skmultiflow.drift_detection.adwin import ADWIN
from skmultiflow.trees.arf_hoeffding_tree import ARFHoeffdingTree

from stream_generator import RecurrentDriftStream
from LRU_state import LRU_state
from state_graph import LossyStateGraph

class Pearl:

    def __init__(self,
                 num_trees,
                 repo_size,
                 edit_distance_threshold,
                 bg_kappa_threshold,
                 cd_kappa_threshold,
                 kappa_window,
                 lossy_window_size,
                 reuse_window_size,
                 reuse_rate_upper_bound,
                 warning_delta,
                 drift_delta,
                 arf_max_features,
                 enable_state_adaption=True,
                 enable_state_graph=True,
                 logger=None):

        self.num_trees = num_trees
        self.warning_delta = warning_delta
        self.drift_delta = drift_delta
        self.repo_size = repo_size
        self.edit_distance_threshold = edit_distance_threshold
        self.bg_kappa_threshold = bg_kappa_threshold
        self.cd_kappa_threshold = cd_kappa_threshold
        self.kappa_window = kappa_window
        self.arf_max_features = arf_max_features

        self.enable_state_adaption = enable_state_adaption
        self.enable_state_graph = enable_state_graph
        self.logger = logger

        self.adaptive_trees = [AdaptiveTree(tree_pool_id=i,
                                            tree=ARFHoeffdingTree(max_features=arf_max_features),
                                            kappa_window=kappa_window,
                                            warning_delta=warning_delta,
                                            drift_delta=drift_delta
                          ) for i in range(0, num_trees)]

        self.cur_state = ['1' if i < num_trees else '0' for i in range(0, repo_size)]

        self.lru_states = LRU_state(capacity=repo_size, edit_distance_threshold=edit_distance_threshold)
        self.lru_states.enqueue(self.cur_state)

        self.state_graph = LossyStateGraph(repo_size, lossy_window_size)

        self.graph_switch = GraphSwitch(window_size=reuse_window_size,
                                   state_graph=self.state_graph,
                                   reuse_rate=reuse_rate_upper_bound)

        self.cur_tree_pool_size = num_trees
        self.candidate_trees = []

        self.tree_pool = [None] * repo_size
        for i in range(0, num_trees):
            self.tree_pool[i] = self.adaptive_trees[i]

    def update_drift_detector(self, adaptive_tree, predicted_label, actual_label):
        if predicted_label == actual_label:
            adaptive_tree.warning_detector.add_element(0)
            adaptive_tree.drift_detector.add_element(0)
        else:
            adaptive_tree.warning_detector.add_element(1)
            adaptive_tree.drift_detector.add_element(1)

    def predict(self, X, y, adaptive_trees, should_vote):
        predictions = []

        for i in range(0, len(X)):
            feature_row = X[i]
            label = y[i]

            votes = defaultdict(int)
            for adaptive_tree in adaptive_trees:
                predicted_label = adaptive_tree.tree.predict([feature_row])[0]

                adaptive_tree.predicted_labels.append(predicted_label) # for kappa calculation
                if should_vote:
                    self.update_drift_detector(adaptive_tree, predicted_label, label)

                votes[predicted_label] += 1

                # background tree needs to predict for performance measurement
                if adaptive_tree.bg_adaptive_tree is not None:
                    self.predict([feature_row], [label], [adaptive_tree.bg_adaptive_tree], False)

            if should_vote:
                predictions.append(max(votes, key=votes.get))

        return predictions

    def partial_fit(self, X, y):
        for i in range(0, len(X)):
            for adaptive_tree in self.adaptive_trees:
                n = np.random.poisson(1)
                adaptive_tree.tree.partial_fit([X[i]], [y[i]], sample_weight=[n])
                if adaptive_tree.bg_adaptive_tree is not None:
                    adaptive_tree.bg_adaptive_tree.tree.partial_fit([X[i]], [y[i]], sample_weight=[n])

    def update_candidate_trees(self,
                               closest_state):
        if len(closest_state) == 0:
            return

        for i in range(0, self.cur_tree_pool_size):

            if self.cur_state[i] == '0' \
                    and closest_state[i] == '1' \
                    and not self.tree_pool[i].is_candidate:

                if len(self.candidate_trees) >= self.num_trees:
                    worst_candidate = self.candidate_trees.pop(0)
                    worst_candidate.reset()

                self.tree_pool[i].is_candidate = True
                self.candidate_trees.append(self.tree_pool[i])

    def select_candidate_trees(self,
                               count,
                               target_state,
                               warning_tree_id_list):

        if self.enable_state_graph:
            # try trigger lossy counting
            if self.state_graph.update(len(warning_tree_id_list)):
                self.logger.info(f"{count},lossy counting triggered")

        if self.state_graph.is_stable:
            for warning_tree_id in warning_tree_id_list:

                next_id = self.state_graph.get_next_tree_id(warning_tree_id)
                if next_id == -1:
                    self.state_graph.is_stable = False

                else:
                    if not self.tree_pool[next_id].is_candidate:
                        self.candidate_trees.append(self.tree_pool[next_id])

        if not self.state_graph.is_stable:
            self.logger.info(f"{count},pattern matching")

            # trigger pattern matching
            closest_state = self.lru_states.get_closest_state(target_state)

            self.update_candidate_trees(closest_state=closest_state)
        else:
            self.logger.info(f"{count},graph transition")

    def adapt_state(self,
                    drifted_tree_list,
                    drifted_tree_pos,
                    actual_labels):

        # sort candidates by kappa
        for candidate_tree in self.candidate_trees:
            candidate_tree.update_kappa(actual_labels)
        self.candidate_trees.sort(key=lambda c : c.kappa)

        for drifted_tree in drifted_tree_list:
            # TODO
            if self.cur_tree_pool_size >= self.repo_size:
                print("early break")
                exit()

            drifted_tree.update_kappa(actual_labels)
            swap_tree = drifted_tree

            background_count = 0
            candidate_count = 0

            if len(self.candidate_trees) > 0 \
                    and self.candidate_trees[-1].kappa - drifted_tree.kappa >= self.cd_kappa_threshold:
                # swap drifted tree with the candidate tree
                swap_tree = self.candidate_trees.pop()
                swap_tree.is_candidate = False

                # candidate_count += 1
                if self.enable_state_graph:
                    self.graph_switch.update(1)

            if swap_tree is drifted_tree:
                add_to_repo = True
                # background_count += 1
                if self.enable_state_graph:
                    self.graph_switch.update(0)

                if drifted_tree.bg_adaptive_tree is None:
                        swap_tree = \
                            AdaptiveTree(tree=ARFHoeffdingTree(max_features=self.arf_max_features),
                                         kappa_window=self.kappa_window,
                                         warning_delta=self.warning_delta,
                                         drift_delta=self.drift_delta)

                else:
                    prediction_win_size = len(drifted_tree.bg_adaptive_tree.predicted_labels)

                    drifted_tree.bg_adaptive_tree.update_kappa(actual_labels)

                    if drifted_tree.bg_adaptive_tree.kappa == -sys.maxsize:
                        # add bg_adaptive_tree to the repo even if it didn't fill the window
                        swap_tree = drifted_tree.bg_adaptive_tree

                    elif drifted_tree.bg_adaptive_tree.kappa - swap_tree.kappa >= self.bg_kappa_threshold:
                        swap_tree = drifted_tree.bg_adaptive_tree

                    else:
                        # false positive
                        add_to_repo = False

                if add_to_repo:
                    swap_tree.reset()

                    # assign a new tree_pool_id for background tree
                    # and add background tree to tree_pool
                    swap_tree.tree_pool_id = self.cur_tree_pool_size
                    self.tree_pool[self.cur_tree_pool_size] = swap_tree
                    self.cur_tree_pool_size += 1

            self.cur_state[drifted_tree.tree_pool_id] = '0'
            self.cur_state[swap_tree.tree_pool_id] = '1'

            if self.enable_state_graph:
                self.state_graph.add_edge(drifted_tree.tree_pool_id, swap_tree.tree_pool_id)

            # replace drifted tree with swap tree
            pos = drifted_tree_pos.pop()
            self.adaptive_trees[pos] = swap_tree
            drifted_tree.reset()

        if self.enable_state_graph:
            self.graph_switch.switch()

    def handle_drift(self, count, actual_labels):

        target_state = copy.deepcopy(self.cur_state)

        warning_tree_id_list = []
        drifted_tree_list = []
        drifted_tree_pos = []

        for i in range(0, self.num_trees):

            tree = self.adaptive_trees[i]
            warning_detected_only = False

            if tree.warning_detector.detected_change():
                warning_detected_only = True
                tree.warning_detector.reset()

                tree.bg_adaptive_tree = \
                    AdaptiveTree(tree=ARFHoeffdingTree(max_features=self.arf_max_features),
                                 kappa_window=self.kappa_window,
                                 warning_delta=self.warning_delta,
                                 drift_delta=self.drift_delta)

            if tree.drift_detector.detected_change():
                warning_detected_only = False
                tree.drift_detector.reset()
                drifted_tree_list.append(tree)
                drifted_tree_pos.append(i)

                if not self.enable_state_adaption:
                    if tree.bg_adaptive_tree is None:
                        tree.tree = ARFHoeffdingTree(max_features=self.arf_max_features)
                    else:
                        tree.tree = tree.bg_adaptive_tree.tree
                    tree.reset()

            if warning_detected_only:
                warning_tree_id_list.append(tree.tree_pool_id)
                target_state[tree.tree_pool_id] = '2'

        if self.enable_state_adaption:
            # if warnings are detected, find closest state and update candidate_trees list
            if len(warning_tree_id_list) > 0:
                self.select_candidate_trees(count=count,
                                            target_state=target_state,
                                            warning_tree_id_list=warning_tree_id_list)

            # if actual drifts are detected, swap trees and update cur_state
            if len(drifted_tree_list) > 0:
                self.adapt_state(drifted_tree_list=drifted_tree_list,
                                 drifted_tree_pos=drifted_tree_pos,
                                 actual_labels=actual_labels)

            self.lru_states.enqueue(self.cur_state)

    def get_size(self):
        memory_usage = 0

        if self.enable_state_adaption:
            memory_usage = self.lru_states.get_size()
        if self.enable_state_graph:
            memory_usage += self.state_graph.get_size()

        return memory_usage


class AdaptiveTree(object):
    def __init__(self,
                 tree,
                 kappa_window,
                 warning_delta,
                 drift_delta,
                 tree_pool_id=-1):
        self.tree_pool_id = tree_pool_id
        self.tree = tree
        self.bg_adaptive_tree = None
        self.is_candidate = False
        self.warning_detector = ADWIN(warning_delta)
        self.drift_detector = ADWIN(drift_delta)
        self.predicted_labels = deque(maxlen=kappa_window)
        self.kappa = -sys.maxsize
        self.kappa_window = kappa_window

    def update_kappa(self, actual_labels):
        if len(self.predicted_labels) < self.kappa_window:
            self.kappa = -sys.maxsize
        else:
            self.kappa = cohen_kappa_score(actual_labels, self.predicted_labels)
        return self.kappa

    def reset(self):
        self.bg_adaptive_tree = None
        self.is_candidate = False
        self.warning_detector.reset()
        self.drift_detector.reset()
        self.predicted_labels.clear()
        self.kappa = -sys.maxsize


class GraphSwitch:
    def __init__(self, window_size, state_graph, reuse_rate):
        self.window_size = window_size
        self.candidate_tree_count = 0
        self.total_tree_count = 0
        self.state_graph = state_graph
        self.reuse_rate = reuse_rate
        if window_size > 0:
            self.window = deque(maxlen=window_size)

    def update(self, value):
        self.candidate_tree_count += value
        self.total_tree_count += 1

        if self.window_size <= 0:
            return

        if len(self.window) >= self.window_size:
            self.candidate_tree_count -= self.window[0]
        self.window.append(value)

    def switch(self):
        cur_reuse_rate = 0
        if self.window_size <= 0:
            cur_reuse_rate = self.candidate_tree_count / self.total_tree_count
        else:
            cur_reuse_rate = self.candidate_tree_count / self.window_size

        if cur_reuse_rate >= self.reuse_rate:
            self.state_graph.is_stable = True
        else:
            self.state_graph.is_stable = False

File: src/cpp/pearl.cpp
#include "pearl.h"

pearl::pearl(int num_trees,
             int max_num_candidate_trees,
             int repo_size,
             int edit_distance_threshold,
             int kappa_window_size,
             int lossy_window_size,
             int reuse_window_size,
             int arf_max_features,
             double bg_kappa_threshold,
             double cd_kappa_threshold,
             double reuse_rate_upper_bound,
             double warning_delta,
             double drift_delta,
             bool enable_state_adaption) :
        num_trees(num_trees),
        max_num_candidate_trees(max_num_candidate_trees),
        repo_size(repo_size),
        edit_distance_threshold(edit_distance_threshold),
        kappa_window_size(kappa_window_size),
        lossy_window_size(lossy_window_size),
        reuse_window_size(reuse_window_size),
        arf_max_features(arf_max_features),
        bg_kappa_threshold(bg_kappa_threshold),
        cd_kappa_threshold(cd_kappa_threshold),
        reuse_rate_upper_bound(reuse_rate_upper_bound),
        warning_delta(warning_delta),
        drift_delta(drift_delta),
        enable_state_adaption(enable_state_adaption) {

    init();
}

void pearl::init() {

    tree_pool = vector<shared_ptr<adaptive_tree>>(num_trees);

    for (int i = 0; i < num_trees; i++) {
        tree_pool[i] = make_adaptive_tree(i);
        adaptive_trees.push_back(tree_pool[i]);
    }

    // initialize LRU state pattern queue
    state_queue = make_unique<lru_state>(10000000, edit_distance_threshold); // TODO

    cur_state = vector<char>(repo_size, '0');
    for (int i = 0; i < num_trees; i++) {
        cur_state[i] = '1';
    }

    state_queue->enqueue(cur_state);

}

shared_ptr<pearl::adaptive_tree> pearl::make_adaptive_tree(int tree_pool_id) {
    return make_shared<adaptive_tree>(tree_pool_id,
                                      kappa_window_size,
                                      warning_delta,
                                      drift_delta);
}

bool pearl::init_data_source(const string& filename) {

    LOG("Initializing data source...");

    reader = make_unique<ArffReader>();

    if (!reader->setFile(filename)) {
        std::cout << "Failed to open file: " << filename << std::endl;
        exit(1);
    }

    return true;
}

void pearl::prepare_instance(Instance& instance) {
    vector<int> attribute_indices;

    // select random features
    for (int i = 0; i < arf_max_features; i++) {
        attribute_indices.push_back(rand() % num_features);
    }

    instance.setAttributeStatus(attribute_indices);
}

bool pearl::process() {
    int actual_label = instance->getLabel();

    int num_classes = instance->getNumberClasses();
    vector<int> votes(num_classes, 0);

    if (enable_state_adaption) {
        process_with_state_adaption(votes, actual_label);

    } else {
        process_basic(votes, actual_label);
    }

    train(*instance);
    delete instance;

    int predicted_label = vote(votes);

    return predicted_label == actual_label;
}

void pearl::process_with_state_adaption(vector<int>& votes, int actual_label) {

    // keep track of actual labels for candidate tree evaluations
    if (actual_labels.size() >= kappa_window_size) {
        actual_labels.pop_front();
    }
    actual_labels.push_back(actual_label);

    int predicted_label;
    vector<int> warning_tree_pos_list;
    vector<int> drifted_tree_pos_list;

    for (int i = 0; i < num_trees; i++) {

        predicted_label = adaptive_trees[i]->predict(*instance, true);

        votes[predicted_label]++;
        int error_count = (int)(actual_label != predicted_label);

        bool warning_detected_only = false;

        // detect warning
        if (detect_change(error_count, adaptive_trees[i]->warning_detector)) {
            warning_detected_only = false;

            adaptive_trees[i]->bg_adaptive_tree = make_adaptive_tree(-1);
            adaptive_trees[i]->warning_detector->resetChange();
        }

        // detect drift
        if (detect_change(error_count, adaptive_trees[i]->drift_detector)) {
            warning_detected_only = true;
            drifted_tree_pos_list.push_back(i);

            adaptive_trees[i]->drift_detector->resetChange();
        }

        if (warning_detected_only) {
            warning_tree_pos_list.push_back(i);
        }
    }

    for (int i = 0; i < candidate_trees.size(); i++) {
        candidate_trees[i]->predict(*instance, true);
    }

    // if warnings are detected, find closest state and update candidate_trees list
    if (warning_tree_pos_list.size() > 0) {
        select_candidate_trees(warning_tree_pos_list);
    }

    // if actual drifts are detected, swap trees and update cur_state
    drift_detected = false;
    if (drifted_tree_pos_list.size() > 0) {
        adapt_state(drifted_tree_pos_list);
        drift_detected = true;
    }
}

void pearl::process_basic(vector<int>& votes, int actual_label) {
    int predicted_label;

    for (int i = 0; i < num_trees; i++) {

        predicted_label = adaptive_trees[i]->predict(*instance, true);

        votes[predicted_label]++;
        int error_count = (int)(actual_label != predicted_label);

        // detect warning
        if (detect_change(error_count, adaptive_trees[i]->warning_detector)) {
            adaptive_trees[i]->bg_adaptive_tree = make_adaptive_tree(-1);
            adaptive_trees[i]->warning_detector->resetChange();
        }

        // detect drift
        if (detect_change(error_count, adaptive_trees[i]->drift_detector)) {
            if (adaptive_trees[i]->bg_adaptive_tree) {
                adaptive_trees[i] = move(adaptive_trees[i]->bg_adaptive_tree);
            } else {
                adaptive_trees[i] = make_adaptive_tree(tree_pool.size());
            }
        }
    }
}

int pearl::vote(vector<int> votes) {
    int max_votes = votes[0];
    int predicted_label = 0;

    for (int i = 1; i < votes.size(); i++) {
        if (max_votes < votes[i]) {
            max_votes = votes[i];
            predicted_label = i;
        }
    }

    return predicted_label;
}

void pearl::train(Instance& instance) {
    for (int i = 0; i < num_trees; i++) {
        // online bagging
        prepare_instance(instance);
        int weight = Utils::poisson(1.0);
        while (weight > 0) {
            weight--;
            adaptive_trees[i]->train(instance);
        }
    }
}

void pearl::select_candidate_trees(vector<int>& warning_tree_pos_list) {

    vector<char> target_state(cur_state);

    for (int i = 0; i < warning_tree_pos_list.size(); i++) {
        int tree_pos = warning_tree_pos_list[i];

        if (adaptive_trees[tree_pos]->tree_pool_id == -1) {
            LOG("Error: tree_pool_id is not updated");
            exit(1);
        }

        target_state[adaptive_trees[tree_pos]->tree_pool_id] = '2';
    }

    vector<char> closest_state = state_queue->get_closest_state(target_state);
    if (closest_state.size() == 0) {
        return;
    }

    for (int i = 0; i < tree_pool.size(); i++) {
        if (cur_state[i] == '0'
            && closest_state[i] == '1'
            && !tree_pool[i]->is_candidate) {

            if (candidate_trees.size() >= max_num_candidate_trees) {
                candidate_trees[0]->is_candidate = false;
                candidate_trees.pop_front();
            }

            tree_pool[i]->is_candidate = true;
            candidate_trees.push_back(tree_pool[i]);
        }
    }
}

void pearl::adapt_state(vector<int> drifted_tree_pos_list) {
    int class_count = instance->getNumberClasses();

    // sort candiate trees by kappa
    for (int i = 0; i < candidate_trees.size(); i++) {
        candidate_trees[i]->update_kappa(actual_labels, class_count);
    }
    sort(candidate_trees.begin(), candidate_trees.end(), compare_kappa);

    for (int i = 0; i < drifted_tree_pos_list.size(); i++) {
        // TODO
        if (tree_pool.size() >= repo_size) {
            std::cout << "tree_pool full: "
                      << std::to_string(tree_pool.size()) << endl;
            exit(1);
        }

        int drifted_pos = drifted_tree_pos_list[i];
        shared_ptr<adaptive_tree> drifted_tree = adaptive_trees[drifted_pos];
        shared_ptr<adaptive_tree> swap_tree;

        drifted_tree->update_kappa(actual_labels, class_count);

        cur_state[drifted_tree->tree_pool_id] = '0';

        bool add_to_repo = false;

        if (candidate_trees.size() > 0
            && candidate_trees.back()->kappa
                - drifted_tree->kappa >= cd_kappa_threshold) {
            candidate_trees.back()->is_candidate = false;
            swap_tree = candidate_trees.back();
            candidate_trees.pop_back();
        }

        if (swap_tree == nullptr) {
            add_to_repo = true;

            shared_ptr<adaptive_tree> bg_tree = drifted_tree->bg_adaptive_tree;

            if (!bg_tree) {
                swap_tree = make_adaptive_tree(tree_pool.size());

            } else {
                bg_tree->update_kappa(actual_labels, class_count);

                if (bg_tree->kappa == INT_MIN) {
                    // add bg tree to the repo even if it didn't fill the window

                } else if (bg_tree->kappa - drifted_tree->kappa >= bg_kappa_threshold) {

                } else {
                    // false positive
                    add_to_repo = false;

                }

                swap_tree = bg_tree;
            }

            if (add_to_repo) {
                swap_tree->reset();

                // assign a new tree_pool_id for background tree
                // and allocate a slot for background tree in tree_pool
                swap_tree->tree_pool_id = tree_pool.size();
                tree_pool.push_back(swap_tree);

            } else {
                swap_tree->tree_pool_id = drifted_tree->tree_pool_id;

                // TODO
                // swap_tree = move(drifted_tree);
            }
        }

        if (!swap_tree) {
            LOG("swap_tree is nullptr");
            exit(1);
        }

        cur_state[swap_tree->tree_pool_id] = '1';

        // replace drifted_tree with swap tree
        adaptive_trees[drifted_pos] = swap_tree;

        drifted_tree->reset();
    }

    state_queue->enqueue(cur_state);
}

bool pearl::detect_change(int error_count,
                          unique_ptr<HT::ADWIN>& detector) {

    double old_error = detector->getEstimation();
    bool error_change = detector->setInput(error_count);

    if (!error_change) {
       return false;
    }

    if (old_error > detector->getEstimation()) {
        // error is decreasing
        return false;
    }

    return true;
}

bool pearl::get_next_instance() {
    if (!reader->hasNextInstance()) {
        return false;
    }

    instance = reader->nextInstance();

    num_features = instance->getNumberInputAttributes();
    arf_max_features = log2(num_features) + 1;

    return true;
}

bool pearl::compare_kappa(shared_ptr<adaptive_tree>& tree1,
                          shared_ptr<adaptive_tree>& tree2) {
    return tree1->kappa < tree2->kappa;
}

int pearl::get_candidate_tree_group_size() const {
    return candidate_trees.size();
}

int pearl::get_tree_pool_size() const {
    return tree_pool.size();
}

// class adaptive_tree
pearl::adaptive_tree::adaptive_tree(int tree_pool_id,
                                    int kappa_window_size,
                                    double warning_delta,
                                    double drift_delta) :
        tree_pool_id(tree_pool_id),
        kappa_window_size(kappa_window_size),
        warning_delta(warning_delta),
        drift_delta(drift_delta) {

    tree = make_unique<HT::HoeffdingTree>();
    warning_detector = make_unique<HT::ADWIN>(warning_delta);
    drift_detector = make_unique<HT::ADWIN>(drift_delta);
}

int pearl::adaptive_tree::predict(Instance& instance, bool track_performance) {
    double numberClasses = instance.getNumberClasses();
    double* classPredictions = tree->getPrediction(instance);
    int result = 0;
    double max = classPredictions[0];

    // Find class label with the highest probability
    for (int i = 1; i < numberClasses; i++) {
        if (max < classPredictions[i]) {
            max = classPredictions[i];
            result = i;
        }
    }

    if (track_performance) {
        if (predicted_labels.size() >= kappa_window_size) {
            predicted_labels.pop_front();
        }
        predicted_labels.push_back(result);

        // the background tree performs prediction for performance eval
        if (bg_adaptive_tree) {
            bg_adaptive_tree->predict(instance, track_performance);
        }
    }

    return result;
}

void pearl::adaptive_tree::train(Instance& instance) {
    tree->train(instance);

    if (bg_adaptive_tree) {
        bg_adaptive_tree->train(instance);
    }
}

void pearl::adaptive_tree::update_kappa(deque<int> actual_labels, int class_count) {

    if (predicted_labels.size() < kappa_window_size) {
        kappa = INT_MIN;
        return;
    }

    int confusion_matrix[class_count][class_count] = {};
    int correct = 0;

    for (int i = 0; i < kappa_window_size; i++) {
        confusion_matrix[actual_labels[i]][predicted_labels[i]]++;
        if (actual_labels[i] == predicted_labels[i]) {
            correct++;
        }
    }

    double accuracy = (double) correct / kappa_window_size;

    kappa = compute_kappa(&(confusion_matrix[0][0]), accuracy, kappa_window_size, class_count);
}

double pearl::adaptive_tree::compute_kappa(int* confusion_matrix,
                                           double accuracy,
                                           int sample_count,
                                           int class_count) {
    // computes the Cohen's kappa coefficient
    double p0 = accuracy;
    double pc = 0.0;
    int row_count = class_count;
    int col_count = class_count;

    for (int i = 0; i < row_count; i++) {
        double row_sum = 0;
        for (int j = 0; j < col_count; j++) {
            row_sum += confusion_matrix[i * col_count + j];
        }

        double col_sum = 0;
        for (int j = 0; j < row_count; j++) {
            col_sum += confusion_matrix[j * row_count + i];
        }

        pc += (row_sum / sample_count) * (col_sum / sample_count);
    }

    if (pc == 1) {
        return 1;
    }

    return (p0 - pc) / (1.0 - pc);
}

void pearl::adaptive_tree::reset() {
    bg_adaptive_tree = nullptr;
    is_candidate = false;
    warning_detector->resetChange();
    drift_detector->resetChange();
    predicted_labels.clear();
    kappa = INT_MIN;
}

bool pearl::get_drift_detected() {
    return drift_detected;
}
